index.js

```diff
@@ -243,9 +271,49 @@ Shopify.prototype.graphql = function graphql(data, variables) {
     parseJson: this.options.parseJson,
     timeout: this.options.timeout,
     responseType: 'json',
-    retry: 0,
     method: 'POST',
-    body: json ? this.options.stringifyJson({ query: data, variables }) : data
+    body: json ? this.options.stringifyJson({ query: data, variables }) : data,
+    retry:
+      this.options.maxRetries > 0
+        ? {
+            limit: this.options.maxRetries,
+            // Don't clamp Shopify retry-after header values too low.
+            maxRetryAfter: Infinity,
+            calculateDelay: (retryObject) => {
+              attemptNumber += 1;
+              return (
+                graphQLRetryMS(
+                  attemptNumber,
+                  this.options.maxRetries,
+                  retryObject.error
+                ) || retryObject.computedValue
+              );
+            }
+          }
+        : 0,
+    hooks: {
+      afterResponse:
+        this.options.maxRetries > 0
+          ? [
+              (response) => {
+                if (response.body && response.body.errors) {
+                  const retryDelay = graphQLRetryMS(
```

Comment:
I think `graphQLRetryMS()` is not needed. If I'm not wrong we can use `maybeRetryMS()` directly instead, and remove `attemptNumber`. `got` will handle the number of retries internally. `maybeRetryMS()` can return `null` if the HTTP response has an invalid `Retry-After` header but I hope such case (`response.body.error` + invalid `Retry-After`) is not possible.

Reply:
I'm also not very happy to call `maybeRetryMS()` twice, here and in `calculateDelay()` to compute the same value twice but I have no better idea at the moment.

Reply:
I mean something like this

```diff
diff --git a/index.js b/index.js
index 10f8500..c2a8610 100644
--- a/index.js
+++ b/index.js
@@ -159,9 +159,17 @@ Shopify.prototype.request = function request(uri, method, key, data, headers) {
             limit: this.options.maxRetries,
             // Don't clamp Shopify retry-after header values too low.
             maxRetryAfter: Infinity,
-            calculateDelay: calculateRESTDelay
+            calculateDelay
           }
         : 0,
+    hooks: {
+      afterResponse: [
+        (res) => {
+          this.updateLimits(res.headers['x-shopify-shop-api-call-limit']);
+          return res;
+        }
+      ]
+    },
     method
   };
 
@@ -173,8 +181,6 @@ Shopify.prototype.request = function request(uri, method, key, data, headers) {
     (res) => {
       const body = res.body;
 
-      this.updateLimits(res.headers['x-shopify-shop-api-call-limit']);
-
       if (res.statusCode === 202 && res.headers['location']) {
         const retryAfter = res.headers['retry-after'] * 1000 || 0;
         const { pathname, search } = url.parse(res.headers['location']);
@@ -262,7 +268,6 @@ Shopify.prototype.graphql = function graphql(data, variables) {
 
   const uri = { pathname, ...this.baseUrl };
   const json = variables !== undefined && variables !== null;
-  let attemptNumber = 0;
   const options = {
     headers: {
       ...this.baseHeaders,
@@ -279,48 +284,44 @@ Shopify.prototype.graphql = function graphql(data, variables) {
             limit: this.options.maxRetries,
             // Don't clamp Shopify retry-after header values too low.
             maxRetryAfter: Infinity,
-            calculateDelay: (retryObject) => {
-              attemptNumber += 1;
-              return (
-                graphQLRetryMS(
-                  attemptNumber,
-                  this.options.maxRetries,
-                  retryObject.error
-                ) || retryObject.computedValue
-              );
-            }
+            calculateDelay
           }
         : 0,
     hooks: {
-      afterResponse:
-        this.options.maxRetries > 0
-          ? [
-              (response) => {
-                if (response.body && response.body.errors) {
-                  const retryDelay = graphQLRetryMS(
-                    attemptNumber,
-                    this.options.maxRetries,
-                    { response }
-                  );
-
-                  if (retryDelay) {
-                    // Make got consider this response errored and retry
-                    throw new Error('GraphQL cost limit exceeded, retrying');
-                  }
-                }
-
-                return response;
-              }
-            ]
-          : []
+      afterResponse: [
+        (response) => {
+          if (response.body.extensions && response.body.extensions.cost) {
+            this.updateGraphqlLimits(response.body.extensions.cost);
+          }
+
+          if (this.options.maxRetries === 0) {
+            return response;
+          }
+
+          if (response.body && response.body.errors) {
+            const retryDelay = maybeRetryMS({ response });
+
+            if (retryDelay) {
+              const first = response.body.errors[0];
+              const err = new Error(first.message);
+
+              err.locations = first.locations;
+              err.path = first.path;
+              err.extensions = first.extensions;
+              err.response = response;
+
+              // Make got consider this response errored and retry
+              throw err;
+            }
+          }
+
+          return response;
+        }
+      ]
     }
   };
 
   return got(uri, options).then((res) => {
-    if (res.body.extensions && res.body.extensions.cost) {
-      this.updateGraphqlLimits(res.body.extensions.cost);
-    }
-
     if (res.body.errors) {
       const first = res.body.errors[0];
       const err = new Error(first.message);
@@ -450,29 +451,13 @@ function maybeRetryMS(error) {
 }
 
 /**
- * Return a duration after which to retry a graphql call, or null if it
- * shouldn't be retried
- *
- * @param {Number} attemptNumber Shopify graphql call attempt no
- * @param {Number} maxRetries Max retries in default option
- * @param {Object} error Error object from got call
- * @return {Boolean | undefined}
- * @private
- */
-function graphQLRetryMS(attemptNumber, maxRetries, error) {
-  if (attemptNumber <= maxRetries) {
-    return maybeRetryMS(error);
-  }
-}
-
-/**
- * Got `calculateDelay` hook function passed when calling the REST API
+ * Got `calculateDelay` hook function.
  *
- * @param {Object} retryObject got's input for the retry logic
+ * @param {Object} retryObject Got's input for the retry logic
  * @return {Number}
  * @private
  */
-function calculateRESTDelay(retryObject) {
+function calculateDelay(retryObject) {
   return maybeRetryMS(retryObject.error) || retryObject.computedValue;
 }
```

Also note the calls to `shopify.updateLimits()` and `shopify.updateGraphqlLimits()` inside the `afterResponse` hook to address https://github.com/MONEI/Shopify-api-node/issues/251#issuecomment-487965284.